### DISCUSSION TITLE
Using Class instead of Struct for Library handle feature

### DISCUSSION THREAD

**Discussion by ingo-h** [pupnp/pupnp#267] [Ideas]
It is thought to use a library handle (https://github.com/pupnp/pupnp/pull/255) that will break the current library interface calls. But it is also a chance to do one step more forward. Because we can use C++ now (https://github.com/pupnp/pupnp/pull/260) it is not a big deal to use a Class specifier instead of a Struct for Library handle. The only difference is the default scope of its members (private for Class, public for Struct). But with a Class you have a better mnemonic‎ to use even that.

For my opinion it simplify things a lot and is easy to expand. For example @whyman suggested
> In terms of usage, I was imagining UpnpInit2 (or 3) to return the handle, rather than having to create one and pass it in https://github.com/pupnp/pupnp/pull/255#issuecomment-799237229

But as @mrjimenez noted that has the problem that it
> Breaks the "return error" pattern of the library, because it would return a pointer and we would have to consider returning NULL an error. https://github.com/pupnp/pupnp/pull/255#issuecomment-799471576

This can easily be solved with the constructor of an interface to handle, for example:

    IUpnpInterface myIfObj("eth0", 51515);

The current library calls can simply be wrapped with methods:

    std::string IF_IPV4 = myIfObj.UpnpGetServerIpAddress();

And yes, this is also reentrant.

What do you think?

**Comment by mrjimenez** [comment]
I think your idea in line with what I did a few days ago in UpnpString.cpp, please take a look. The thing is, the old "C" pointer is now a pointer to a class.

You are right, now that we have C++ we should be using classes, but we can only do this internally. To preserve the external C usability, the object pointer must be explicitly passed in the API. Since it is an opaque handler, the C caller cannot do anything with it, unless it uses libupnp routines.

Of course, we can add a C++ interface to the library, but that is something else.

So, your IUpnpInterface is in fact LibUpnp, rewritten as a class, do you agree? That is how I imagined the way forward.

Using UpnpString as an inspiration, it got such an obvious implementation that instead of
```C
class UpnpString : public std::string
{
};
```
I should have done
```C
typedef std::string UpnpString
```
The only reason I did not do that is that the C interface might still be convenient to use in UpnpString_casecmp(). Which, btw, will solve some #ifdefs on WIN32 because of strcasecmp.

The same way, we could remove UpnpLib from the generator and implement it by hand as a class. Or maybe write a generator2 to write the bureaucratic getter/setter methods.
